### PR TITLE
[C] Add rocksdb_options_set_cf_paths

### PR DESCRIPTION
**Pull request by yhchiang-sol**
This PR adds a missing set function for rocksdb_options in the C-API:
rocksdb_options_set_cf_paths().  Without this function, users cannot
specify different paths for different column families as it will fall back
to db_paths.

As a bonus, this PR also includes rocksdb_sst_file_metadata_get_directory()
to the C api -- a missing public function that will also make the test easier to write.

### Test Plan
Augment existing c_test to verify the specified cf_path.

### CODE REVIEW

File: db/c_test.c [hunk 1]
```diff
@@ -1441,10 +1441,20 @@ int main(int argc, char** argv) {
...
+    rocksdb_options_t* cf_options_2 = rocksdb_options_create();
+
+    // use dbpathname2 as the cf_path for "cf1"
+    rocksdb_dbpath_t* dbpath2;
+    char dbpathname2[200];
+    snprintf(dbpathname2, sizeof(dbpathname2), "%s/rocksdb_c_test-%d-dbpath2",
+             GetTempDir(), ((int)geteuid()));
+    dbpath2 = rocksdb_dbpath_create(dbpathname2, 1024 * 1024);
+    const rocksdb_dbpath_t* cf_paths[1] = {dbpath2};
+    rocksdb_options_set_cf_paths(cf_options_2, cf_paths, 1);
```
Comment (yhchiang-sol) [inline db/c_test.c:1454]: Verified the test will fail if we use the existing `rocksdb_options_set_db_paths`.

**Comment by YuraKotov** [comment]
Could you please comment would it fix the case if I setup a DB with multiple CFs via its descriptors?
I tried to plase CFs data to different places by the following Rust code, but it seems it doesnt work:
``` rust
// Create cf1 CF descriptor
let mut cf1_opts = Options::default();
cf1_opts.create_if_missing(true);
let cf1_path = Path::new("").join(&file_path).join("/").join(CF_NAME_CF1);
let cf1_dbpath = rocksdb::DBPath::new(cf1_path, 0).unwrap();
cf1_opts.set_db_paths(&[cf1_dbpath]);
let cf1_desc = rocksdb::ColumnFamilyDescriptor::new(CF_NAME_CF1, cf1_opts);

// Create cf2 CF descriptor
let mut cf2_opts = Options::default();
cf1_opts.create_if_missing(true);
let cf2_path = Path::new("").join(&file_path).join("/").join(CF_NAME_CF2);
let cf2_dbpath = rocksdb::DBPath::new(cf2_path, 0).unwrap();
cf2_opts.set_db_paths(&[cf2_dbpath]);
let cf2_desc = rocksdb::ColumnFamilyDescriptor::new(CF_NAME_CF2, cf2_opts);

let db = TransactionDB::open_cf_descriptors(
	&options,
	&TransactionDBOptions::default(),
	file_path,
	[cf1_desc, cf2_desc],
)
.unwrap();
```

**Comment by yhchiang-sol** [comment]
> What is confusing me is that open_cf_descriptors should work right now and it should be possible to setup different pathes for CF using open_cf_descriptors API.
So, your change looks like an alternative way to do that.
Or I miss some point?

I think what you miss is that `db_paths` is a single global setting that is shared across multiple column families.
In other words, you will not be able to specify different paths for different column families using `db_paths`.

Instead, `cf_paths` should be used, and it's not currently available in the C API before this PR.

Another thing to mention is that `ColumnFamilyOptions` is ported via `Options` in C API instead,
and `Options` is actually `DBOptions` + `ColumnFamilyOptions`,
and `db_paths` actually belongs to `DBOptions`.

So what happens in the C API is that it creates an `Options` when you want to create a `ColumnFamilyOptions`,
and then there's API to specify the DBOptions (in your case it is `db_path`).  So when you use `Options` to 
create a ColumnFamily, those settings in DBOptions are ignored.

**Comment by yhchiang-sol** [comment]
The test failure in `ci/circleci: build-macos-cmake-1` doesn't look like something related to this PR:

```
The following tests FAILED:
	5980 - DBWriteTestInstance/DBWriteTest.LockWALInEffect/29 (SEGFAULT)
```

**Comment by yhchiang-sol** [comment]
Hey, @siying, @ajkr.  I have a quick PR that allows RocksDB C API to specify cf_paths (yep, this basic feature currently missing in the C API!)

The test failure in `ci/circleci: build-macos-cmake-1` doesn't look like something related to this PR.  Can you issue a rerun for me as the rerun button is disabled on my side?

**Review by ajkr — APPROVED** [review]
Thanks!

**Pull request closed without merging**